File: tests/python/test_vtx.py
"""DOLFINx VTX output (ADIOS2 ``.bp`` directories): real DOLFINx 0.11 runs
checked against ParaView 6.1's ``ADIOS2VTXReader`` (and, for the steps ParaView
cannot read, against the raw ADIOS2 arrays); see ``meshes/vtx/README.md``.

The engine is the core when it was built with ADIOS2, else the ``adios2``
package: one process never holds both, since the package brings its own
ADIOS2 libraries."""

import pathlib
import shutil
import sys

import numpy as np
import pytest

import meshioplusplus
from meshioplusplus import _core
from meshioplusplus._exceptions import ReadError
from meshioplusplus.vtx import _vtx

HAS_CORE = getattr(_core, "__has_adios2__", False)
if not HAS_CORE:
    pytest.importorskip("adios2")

VTX = pathlib.Path(__file__).parent / "meshes" / "vtx"
NAMES = ["heat", "elastic", "quads", "heat_np2", "complex"]

# VTK_LAGRANGE_* (type, nodes) -> meshio++ type, as the reader lowers them.
LOWERED = {
    (69, 3): "triangle",
    (70, 4): "quad",
    (71, 10): "tetra10",
    (5, 3): "triangle",
}


def _reference(name):
    return np.load(VTX / f"{name}.reference.npz")


def _raw(name):
    return np.load(VTX / f"{name}.raw.npz")


def _connectivity(mesh):
    return np.concatenate([c.data.ravel() for c in mesh.cells])


@pytest.mark.parametrize("name", NAMES)
def test_matches_paraview(name):
    ref = _reference(name)
    assert meshioplusplus.vtx.time_values(VTX / f"{name}.bp") == list(ref["times"])
    for k in ref["steps"]:
        mesh = meshioplusplus.read(VTX / f"{name}.bp", time_step=int(k))
        np.testing.assert_array_equal(mesh.points, ref[f"k{k}/points"])
        np.testing.assert_array_equal(_connectivity(mesh), ref[f"k{k}/connectivity"])
        sizes = np.diff(ref[f"k{k}/offsets"])
        types = [LOWERED[(t, n)] for t, n in zip(ref[f"k{k}/types"], sizes)]
        assert [c.type for c in mesh.cells] == sorted(set(types), key=types.index)
        assert mesh.field_data["meshio:time"][0] == ref["times"][k]
        for key in ref.files:
            if not key.startswith(f"k{k}/point/") and not key.startswith(f"k{k}/cell/"):
                continue
            kind, array = key.split("/")[1:]
            expected = ref[key]
            if kind == "point":
                got = mesh.point_data[array]
                # ParaView appends DOLFINx's stray one-value ghost blocks.
                expected = expected[: len(got)]
            else:
                got = np.concatenate(mesh.cell_data[array])
            np.testing.assert_array_equal(got, expected)


@pytest.mark.parametrize("name", ["heat", "heat_np2", "elastic", "complex"])
def test_every_step_matches_raw_arrays(name):
    """Every step, the ones ParaView aborts on included: the mesh carried from
    step 0 (`reuse`), the functions as ADIOS2 stored them."""
    raw = _raw(name)
    first = meshioplusplus.read(VTX / f"{name}.bp")
    steps = sorted({int(key.split("/")[0][1:]) for key in raw.files})
    for k in steps:
        mesh = meshioplusplus.read(VTX / f"{name}.bp", time_step=k)
        np.testing.assert_array_equal(mesh.points, first.points)
        np.testing.assert_array_equal(_connectivity(mesh), _connectivity(first))
        for key in raw.files:
            if not key.startswith(f"k{k}/"):
                continue
            array = key.split("/")[1]
            got = (
                mesh.point_data[array]
                if array in mesh.point_data
                else np.concatenate(mesh.cell_data[array])
            )
            np.testing.assert_array_equal(got, raw[key])


def test_ghost_arrays_skip_dolfinx_stray_blocks():
    """The first step of a `reuse` file holds malformed one-value ghost blocks;
    the arrays match the rank blocks and every step's are the same."""
    ids = [
        meshioplusplus.read(VTX / "heat_np2.bp", time_step=k).point_data[
            "vtkOriginalPointIds"
        ]
        for k in range(3)
    ]
    assert len(ids[0]) == 38
    for other in ids[1:]:
        np.testing.assert_array_equal(ids[0], other)
    assert set(ids[0]) == set(range(25))


def _by_position(mesh):
    return {tuple(np.round(p, 9)): i for i, p in enumerate(mesh.points)}


@pytest.mark.parametrize("step", [0, 2])
def test_drop_ghosts_gives_the_serial_mesh(step):
    parallel = meshioplusplus.vtx.read(
        VTX / "heat_np2.bp", time_step=step, ghosts="drop"
    )
    serial = meshioplusplus.read(VTX / "heat.bp", time_step=step)
    assert len(parallel.points) == len(serial.points)
    assert "vtkGhostType" not in parallel.point_data
    index = _by_position(parallel)
    order = np.array([index[tuple(np.round(p, 9))] for p in serial.points])
    np.testing.assert_allclose(
        parallel.point_data["u"][order], serial.point_data["u"], atol=1e-14
    )

    def cells(mesh):
        return {
            tuple(sorted(tuple(np.round(mesh.points[i], 9)) for i in c))
            for c in mesh.cells[0].data
        }

    assert cells(parallel) == cells(serial)


def test_points_only_and_array_filter():
    mesh = meshioplusplus.vtx.read(VTX / "elastic.bp", points_only=True)
    assert not mesh.point_data and not mesh.cell_data
    mesh = meshioplusplus.vtx.read(VTX / "elastic.bp", arrays=["density"])
    assert list(mesh.cell_data) == ["density"] and not mesh.point_data


def test_negative_and_out_of_range_steps():
    last = meshioplusplus.read(VTX / "heat.bp", time_step=-1)
    assert last.field_data["meshio:time"][0] == pytest.approx(0.02)
    with pytest.raises(ReadError, match="3"):
        meshioplusplus.read(VTX / "heat.bp", time_step=3)


def test_metadata_and_sequence():
    meta = meshioplusplus.read_metadata(VTX / "heat.bp")
    assert meta["time_values"] == pytest.approx([0.0, 0.01, 0.02])
    steps = list(meshioplusplus.read_sequence(VTX / "heat.bp"))
    assert [t for t, _ in steps] == pytest.approx([0.0, 0.01, 0.02])


def test_glob_keeps_bp_directories(tmp_path):
    for k in range(2):
        shutil.copytree(VTX / "quads.bp", tmp_path / f"run_{k}.bp")
    (tmp_path / "ignored").mkdir()
    entries = meshioplusplus.sequence_entries(str(tmp_path / "run_*.bp"))
    assert [pathlib.Path(e["path"]).name for e in entries] == ["run_0.bp", "run_1.bp"]


def test_sniffed_as_a_directory(tmp_path):
    target = tmp_path / "noext"
    shutil.copytree(VTX / "heat.bp", target)
    from meshioplusplus._sniff import sniff_format

    assert sniff_format(target) == "vtx"


def test_refuses_a_bp_without_schema():
    with pytest.raises(ReadError, match="adios4dolfinx"):
        meshioplusplus.read(VTX / "plain.bp")


def test_python_twin_names_its_dependency(monkeypatch):
    monkeypatch.setitem(sys.modules, "adios2", None)
    with pytest.raises(ImportError, match=r"meshioplusplus\[adios2\]"):
        _vtx.read(VTX / "heat.bp")

File: tests/cpp/test_vtx.cpp
//  ██████   ██████ ██████████  █████████  █████   █████ █████    ███████
// ░░██████ ██████ ░░███░░░░░█ ███░░░░░███░░███   ░░███ ░░███   ███░░░░░███      ███         ███
//  ░███░█████░███  ░███  █ ░ ░███    ░░░  ░███    ░███  ░███  ███     ░░███    ░███        ░███
//  ░███░░███ ░███  ░██████   ░░█████████  ░███████████  ░███ ░███      ░███ ███████████ ███████████
//  ░███ ░░░  ░███  ░███░░█    ░░░░░░░░███ ░███░░░░░███  ░███ ░███      ░███░░░░░███░░░ ░░░░░███░░░
//  ░███      ░███  ░███ ░   █ ███    ░███ ░███    ░███  ░███ ░░███     ███     ░███        ░███
//  █████     █████ ██████████░░█████████  █████   █████ █████ ░░░███████░      ░░░         ░░░
// ░░░░░     ░░░░░ ░░░░░░░░░░  ░░░░░░░░░  ░░░░░   ░░░░░ ░░░░░    ░░░░░░░
//
//
//  License:         MIT License
//                   meshio++ default license: LICENSE
//
//  Main authors:    Vicente Mataix Ferrandiz
//
//

// DOLFINx VTX (.bp) reader: synthetic files written through ADIOS2's C++ API in
// VTXWriter's layout (the real DOLFINx runs are the Python tests' fixtures).
#ifdef MESHIOPLUSPLUS_HAS_ADIOS2

// System includes
#include <array>
#include <atomic>
#include <cstdint>
#include <filesystem>
#include <string>
#include <vector>

// External includes
#include <adios2.h>
#include <gtest/gtest.h>

// Project includes
#include "meshioplusplus/exceptions.hpp"
#include "meshioplusplus/formats/vtx.hpp"
#include "meshioplusplus/operations/sequence.hpp"
#include "meshioplusplus/operations/sniff.hpp"
#include "meshioplusplus/registry.hpp"

namespace {

using meshioplusplus::GhostPolicy;
using meshioplusplus::Mesh;
using meshioplusplus::ReadError;
using meshioplusplus::ReadOptions;

class VtxTempDir {
public:
    VtxTempDir() {
        static std::atomic<unsigned> counter{0};
        mPath = std::filesystem::temp_directory_path() /
                ("meshio_vtx_" + std::to_string(counter++) + "_" +
                 std::to_string(::testing::UnitTest::GetInstance()->random_seed()));
        std::filesystem::create_directories(mPath);
    }
    ~VtxTempDir() {
        std::error_code ec;
        std::filesystem::remove_all(mPath, ec);
    }
    VtxTempDir(const VtxTempDir&) = delete;
    VtxTempDir& operator=(const VtxTempDir&) = delete;
    std::string operator/(const std::string& rName) const { return (mPath / rName).string(); }
    const std::filesystem::path& Path() const { return mPath; }

private:
    std::filesystem::path mPath;
};

const char* const kSchema = R"(<VTKFile type="UnstructuredGrid" version="0.1">
  <UnstructuredGrid>
    <Piece NumberOfPoints="NumberOfNodes" NumberOfCells="NumberOfCells">
      <Points><DataArray Name="geometry" /></Points>
      <Cells>
        <DataArray Name="connectivity" />
        <DataArray Name="types" />
      </Cells>
      <PointData>
        <DataArray Name="TIME">step</DataArray>
        <DataArray Name="vtkOriginalPointIds" />
        <DataArray Name="vtkGhostType" />
        <DataArray Name="u" />
      </PointData>
      <CellData><DataArray Name="rank" /></CellData>
    </Piece>
  </UnstructuredGrid>
</VTKFile>)";

// One "rank": a local block. Two unit tetrahedra of degree 2 split over two
// ranks that share the face (1, 2, 3): each rank holds its tetrahedron's ten
// nodes, the shared face's six as ghosts on rank 1.
struct Rank {
    std::vector<double> mX;            // n x 3
    std::vector<std::int64_t> mConn;   // 1 x 11, count-prefixed
    std::vector<std::int64_t> mIds;    // global node ids
    std::vector<std::uint8_t> mGhost;  // 1 = ghost
};

// The ten P2 nodes of the tetrahedron (a, b, c, d) in VTK order: corners, then
// the edge midpoints (ab, bc, ca, ad, bd, cd).
std::vector<std::array<double, 3>> vtx_p2_nodes(const std::array<std::array<double, 3>, 4>& rV) {
    auto mid = [&](int i, int j) {
        return std::array<double, 3>{(rV[i][0] + rV[j][0]) / 2, (rV[i][1] + rV[j][1]) / 2,
                                     (rV[i][2] + rV[j][2]) / 2};
    };
    return {rV[0],     rV[1],     rV[2],     rV[3],     mid(0, 1),
            mid(1, 2), mid(2, 0), mid(0, 3), mid(1, 3), mid(2, 3)};
}

std::vector<Rank> vtx_two_ranks() {
    const std::array<double, 3> p0{0, 0, 0}, p1{1, 0, 0}, p2{0, 1, 0}, p3{0, 0, 1}, p4{1, 1, 1};
    const auto a = vtx_p2_nodes({p0, p1, p2, p3});
    const auto b = vtx_p2_nodes({p4, p1, p2, p3});
    // Global ids: the first tetrahedron's nodes 0..9; the second's corner p4 is
    // 10, its edges (p4,p1) 11, (p1,p2) 5, (p2,p4) 12, (p4,p3) 13, (p1,p3) 8,
    // (p2,p3) 9 -- node k of the second tetrahedron in VTK order:
    const std::vector<std::int64_t> ids_b = {10, 1, 2, 3, 11, 5, 12, 13, 8, 9};
    const std::vector<std::uint8_t> ghost_b = {0, 1, 1, 1, 0, 1, 0, 0, 1, 1};
    Rank r0, r1;
    for (int k = 0; k < 10; ++k) {
        r0.mX.insert(r0.mX.end(), a[k].begin(), a[k].end());
        r1.mX.insert(r1.mX.end(), b[k].begin(), b[k].end());
        r0.mIds.push_back(k);
        r0.mGhost.push_back(0);
    }
    r1.mIds = ids_b;
    r1.mGhost = ghost_b;
    r0.mConn.push_back(10);
    r1.mConn.push_back(10);
    for (int k = 0; k < 10; ++k) {
        r0.mConn.push_back(k);
        r1.mConn.push_back(k);
    }
    return {r0, r1};
}

// Writes a VTX file: the mesh in step 0 only (VTXMeshPolicy.reuse), `u` (the
// global id plus 100 * step) and the ghost arrays every step. With
// `StrayBlock`, step 0's `vtkGhostType` gets a trailing one-value block.
void vtx_write(const std::string& rPath, int Steps, bool Schema = true, bool StrayBlock = false) {
    adios2::ADIOS adios;
    adios2::IO io = adios.DeclareIO("vtx_test_writer");
    io.SetEngine("BP5");
    adios2::Engine engine = io.Open(rPath, adios2::Mode::Write);
    if (Schema)
        io.DefineAttribute<std::string>("vtk.xml", kSchema);
    const std::vector<Rank> ranks = vtx_two_ranks();
    auto step_var = io.DefineVariable<double>("step");
    auto geometry = io.DefineVariable<double>("geometry", {}, {}, {10, 3});
    auto conn = io.DefineVariable<std::int64_t>("connectivity", {}, {}, {1, 11});
    auto types = io.DefineVariable<std::uint32_t>("types");
    auto ids = io.DefineVariable<std::int64_t>("vtkOriginalPointIds", {}, {}, {10});
    auto ghost = io.DefineVariable<std::uint8_t>("vtkGhostType", {}, {}, {10});
    auto u = io.DefineVariable<double>("u", {}, {}, {10});
    auto rank_var = io.DefineVariable<double>("rank", {}, {}, {1});
    for (int s = 0; s < Steps; ++s) {
        engine.BeginStep();
        engine.Put(step_var, 0.5 * s);
        if (s == 0)
            engine.Put(types, std::uint32_t{71});
        for (std::size_t r = 0; r < ranks.size(); ++r) {
            const Rank& rR = ranks[r];
            if (s == 0) {
                engine.Put(geometry, rR.mX.data(), adios2::Mode::Sync);
                engine.Put(conn, rR.mConn.data(), adios2::Mode::Sync);
                const double rv = static_cast<double>(r);
                engine.Put(rank_var, &rv, adios2::Mode::Sync);
            }
            engine.Put(ids, rR.mIds.data(), adios2::Mode::Sync);
            engine.Put(ghost, rR.mGhost.data(), adios2::Mode::Sync);
            std::vector<double> values;
            for (std::int64_t id : rR.mIds)
                values.push_back(static_cast<double>(id) + 100.0 * s);
            engine.Put(u, values.data(), adios2::Mode::Sync);
        }
        if (StrayBlock && s == 0) {
            ghost.SetSelection({{}, {1}});
            const std::uint8_t one = 1;
            engine.Put(ghost, &one, adios2::Mode::Sync);
            ghost.SetSelection({{}, {10}});
        }
        engine.EndStep();
    }
    engine.Close();
}

TEST(Vtx, ConcatenatesRankBlocksAndLowersLagrangeCells) {
    VtxTempDir dir;
    const std::string path = dir / "run.bp";
    vtx_write(path, 3);
    const Mesh mesh = meshioplusplus::read_vtx(path);
    ASSERT_EQ(mesh.NumPoints(), 20u);
    ASSERT_EQ(mesh.NumCellBlocks(), 1u);
    const auto cells = mesh.Cells(0);
    EXPECT_EQ(cells.Type(), "tetra10");
    ASSERT_EQ(cells.NumCells(), 2u);
    const auto* c = cells.Conn().As<std::int64_t>();
    EXPECT_EQ(c[0], 0);
    EXPECT_EQ(c[10], 10);  // the second rank's numbering is offset
    EXPECT_EQ(c[19], 19);
    const auto* rank = mesh.CellData("rank", 0).As<double>();
    EXPECT_EQ(rank[0], 0.0);
    EXPECT_EQ(rank[1], 1.0);
    EXPECT_EQ(mesh.FieldData("meshio:time").As<double>()[0], 0.0);
}

TEST(Vtx, ReusePolicyCarriesTheMeshForward) {
    VtxTempDir dir;
    const std::string path = dir / "run.bp";
    vtx_write(path, 3);
    ReadOptions opts;
    opts.mTimeStep = 2;
    const Mesh mesh = meshioplusplus::read_vtx(path, opts);
    EXPECT_EQ(mesh.NumPoints(), 20u);
    EXPECT_EQ(mesh.FieldData("meshio:time").As<double>()[0], 1.0);
    const auto* u = mesh.PointData("u").As<double>();
    EXPECT_EQ(u[0], 200.0);
    EXPECT_EQ(u[10], 210.0);
    EXPECT_EQ(meshioplusplus::vtx_time_values(path), (std::vector<double>{0.0, 0.5, 1.0}));
    opts.mTimeStep = 3;
    EXPECT_THROW(meshioplusplus::read_vtx(path, opts), ReadError);
}

TEST(Vtx, DropGhostsWeldsOntoOwners) {
    VtxTempDir dir;
    const std::string path = dir / "run.bp";
    vtx_write(path, 1);
    ReadOptions opts;
    opts.mGhosts = GhostPolicy::Drop;
    const Mesh mesh = meshioplusplus::read_vtx(path, opts);
    EXPECT_EQ(mesh.NumPoints(), 14u);  // 20 minus the shared face's six ghosts
    EXPECT_FALSE(mesh.HasPointData("vtkGhostType"));
    EXPECT_FALSE(mesh.HasPointData("vtkOriginalPointIds"));
    const auto* c = mesh.Cells(0).Conn().As<std::int64_t>();
    const auto* u = mesh.PointData("u").As<double>();
    // The second tetrahedron's corner 1 is the first one's corner 1 (id 1).
    EXPECT_EQ(c[11], c[1]);
    EXPECT_EQ(u[c[10]], 10.0);
}

TEST(Vtx, SkipsBlocksMatchingNoRank) {
    VtxTempDir dir;
    const std::string path = dir / "run.bp";
    vtx_write(path, 2, true, true);
    const Mesh mesh = meshioplusplus::read_vtx(path);
    ASSERT_TRUE(mesh.HasPointData("vtkGhostType"));
    EXPECT_EQ(mesh.PointData("vtkGhostType").Shape()[0], 20u);
}

TEST(Vtx, RefusesAFileWithoutSchema) {
    VtxTempDir dir;
    const std::string path = dir / "plain.bp";
    vtx_write(path, 1, false);
    try {
        meshioplusplus::read_vtx(path);
        FAIL() << "expected a ReadError";
    } catch (const ReadError& rE) {
        EXPECT_NE(std::string(rE.what()).find("adios4dolfinx"), std::string::npos);
    }
    EXPECT_THROW(meshioplusplus::read_vtx(dir / "missing.bp"), ReadError);
}

TEST(Vtx, MetadataFromTheBlockLists) {
    VtxTempDir dir;
    const std::string path = dir / "run.bp";
    vtx_write(path, 3);
    const auto meta = meshioplusplus::read_vtx_metadata(path, {});
    EXPECT_FALSE(meta.mFellBackToFullRead);
    EXPECT_EQ(meta.mNumPoints, 20u);
    ASSERT_EQ(meta.mCellBlocks.size(), 1u);
    EXPECT_EQ(meta.mCellBlocks[0].mType, "tetra10");
    EXPECT_EQ(meta.mCellBlocks[0].mNumCells, 2u);
    EXPECT_EQ(meta.mTimeValues, (std::vector<double>{0.0, 0.5, 1.0}));
    EXPECT_EQ(meta.mPointDataNames,
              (std::vector<std::string>{"u", "vtkGhostType", "vtkOriginalPointIds"}));
    EXPECT_EQ(meta.mCellDataNames, (std::vector<std::string>{"rank"}));
}

TEST(Vtx, RegistrySniffAndSequenceGlob) {
    VtxTempDir dir;
    vtx_write(dir / "run_0.bp", 1);
    vtx_write(dir / "run_1.bp", 2);
    EXPECT_EQ(meshioplusplus::resolve_format(dir / "run_0.bp", ""), "vtx");
    std::filesystem::rename(dir / "run_1.bp", dir / "noext");
    EXPECT_EQ(meshioplusplus::sniff_format(dir / "noext"), "vtx");
    std::filesystem::rename(dir / "noext", dir / "run_1.bp");
    EXPECT_TRUE(meshioplusplus::seq_format_may_have_steps("vtx"));
    meshioplusplus::SequenceInput in;
    in.mPattern = dir / "run_*.bp";
    const auto entries = meshioplusplus::sequence_expand(in);
    ASSERT_EQ(entries.size(), 3u);  // one step, then two
    EXPECT_EQ(std::filesystem::path(entries[0].mPath).filename(), "run_0.bp");
    EXPECT_EQ(std::filesystem::path(entries[2].mPath).filename(), "run_1.bp");
}

}  // namespace

#endif  // MESHIOPLUSPLUS_HAS_ADIOS2
